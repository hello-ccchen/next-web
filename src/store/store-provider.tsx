"use client";
import { Provider } from "react-redux";
import { CommonProps } from "@/interfaces/common/common-interface";
import { store } from "./store";

function StoreProvider({ children }: CommonProps) {
  return <Provider store={store}>{children}</Provider>;
}

export default StoreProvider;
