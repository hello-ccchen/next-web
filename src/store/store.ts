import { configureStore } from "@reduxjs/toolkit";
import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";
import usersReducer from "@/store/users/users-slice";
import productsReducer from "@/store/products/products-slice";

export const store = configureStore({
  reducer: {
    users: usersReducer,
    products: productsReducer
  },
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
