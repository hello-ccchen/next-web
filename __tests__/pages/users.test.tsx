/**
 * @jest-environment jsdom
 */
import { render } from "@testing-library/react";
import Users from "@/app/users/page";
import * as reactRedux from "@/store/store";
import { mockUserResponseDataWithOnePage } from "../config/mocks/users-data.mock";

describe("Users Page", () => {
  const useAppSelectorMock = jest.spyOn(reactRedux, "useAppSelector");
  const useAppDispatchMock = jest.spyOn(reactRedux, "useAppDispatch");

  afterEach(() => {
    useAppSelectorMock.mockClear();
    useAppDispatchMock.mockClear();
  });

  it("should renders correctly when contain records", () => {
    // Arrange
    useAppSelectorMock.mockReturnValue({
      userList: mockUserResponseDataWithOnePage.data,
      isLoading: false,
      error: undefined,
    });
    useAppDispatchMock.mockReturnValue(jest.fn());

    // Act
    const { container } = render(<Users />);

    // Assert
    expect(container).toMatchSnapshot();
    const divElements = container.querySelectorAll(".col-sm-4");
    expect(divElements).toBeTruthy();
    expect(divElements.length).toBe(
      mockUserResponseDataWithOnePage.data.length
    );
  });

  it("should renders `Loading...` when loading is true", () => {
    // Arrange
    useAppSelectorMock.mockReturnValue({
      userList: [],
      isLoading: true,
      error: undefined,
    });
    useAppDispatchMock.mockReturnValue(jest.fn());

    // Act
    const { container, getByText } = render(<Users />);

    // Assert
    expect(container).toMatchSnapshot();
    expect(getByText("Loading...")).toBeTruthy();
  });

  it("should renders `No data` when contain no records", () => {
    // Arrange
    useAppSelectorMock.mockReturnValue({
      userList: [],
      isLoading: false,
      error: undefined,
    });
    useAppDispatchMock.mockReturnValue(jest.fn());

    // Act
    const { container, getByText } = render(<Users />);

    // Assert
    expect(container).toMatchSnapshot();
    expect(getByText("Uh-oh, no data")).toBeTruthy();
  });
});
