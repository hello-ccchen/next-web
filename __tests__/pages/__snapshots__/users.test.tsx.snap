// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users Page should renders \`Loading...\` when loading is true 1`] = `
<div>
  <h1>
    Loading...
  </h1>
</div>
`;

exports[`Users Page should renders \`No data\` when contain no records 1`] = `
<div>
  <h1>
    Uh-oh, no data
  </h1>
</div>
`;

exports[`Users Page should renders correctly when contain records 1`] = `
<div>
  <div
    class="row mx-auto w-75 container"
  >
    <div
      class="col-sm-4"
    >
      <div
        class="d-flex flex-wrap justify-content-center align-items-center my-auto"
      >
        <img
          alt="avatar"
          class="m-2 rounded-circle"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="150"
          src="/_next/image?url=https%3A%2F%2Fexample%2Fguava.jpg&w=384&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fexample%2Fguava.jpg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fexample%2Fguava.jpg&w=384&q=75 2x"
          style="color: transparent;"
          width="150"
        />
        <div
          class="d-flex flex-column align-items-center w-100"
        >
          <h6>
            Guava Banana
          </h6>
          <div
            class="d-flex align-items-center"
          >
            <span>
              *****
            </span>
            <img
              alt="mask"
              class="ms-2"
              data-nimg="1"
              decoding="async"
              height="16"
              loading="lazy"
              src="/eye.svg"
              style="color: transparent; cursor: pointer;"
              width="16"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
